docs/design-docs/05-kpt-fn-exclude.md

```diff
@@ -0,0 +1,233 @@
+
+A few reasons in favor of this design:
+- Flexible and powerful
+- Allows users very explicit control of which resources are included at a per-function level
+- It is consistent with our current selector mechanism.
+- It is consistent with the syntax of kustomize replacements (used in our new apply-replacements function), which is popular and heavily used among the kubernetes community
+- It is consistent with a highly requested feature in kustomize to allow transformers to exclude resources of a certain GVK.
+
+#### Example: Exclude all resources of kind "Deployment"
+
+```shell
+$ kpt fn eval [PKG_DIR] -i set-labels:v0.1.5 --exclude-kind Deployment
+```
+
+```yaml
+# pipeline of Kptfile
+pipeline:
+  mutators:
+  - image: set-labels:v0.1.5
+    exclude:
+    - kind: Deployment
+```
+
+#### Example: Exclude all resources that have both group "apps" and kind "Deployment"
+
+```shell
+$ kpt fn eval [PKG_DIR] -i set-labels:v0.1.5 --exclude-kind Deployment --exclude-group apps
+```
+```yaml
+# pipeline of Kptfile
+pipeline:
+  mutators:
+  - image: set-labels:v0.1.5
+    exclude:
+    - kind: Deployment
+      group: apps
+```
+
+#### Example: Exclude all resources that have either group "apps" or kind "Deployment"
+
+```yaml
+# pipeline of Kptfile
+pipeline:
+  mutators:
+  - image: set-labels:v0.1.5
+    exclude:
+    - kind: Deployment
+    - group: apps
+```
+
+With the current proposal, this is not possible imperatively with kpt fn eval.
+
+#### Example: Select all resources that have group "apps", but do NOT have kind "Deployment"
+
+```shell
+$ kpt fn eval [PKG_DIR] -i set-labels:v0.1.5 --match-group apps --exclude-kind Deployment
+```
+```yaml
+# pipeline of Kptfile
+pipeline:
+  mutators:
+  - image: set-labels:v0.1.5
+    selectors:
+      group: apps
+    exclude:
+    - kind: Deployment
+``` 
+
+### Selecting by labels or annotations
+
+We will support selection and exclusion with annotation and label selectors.
+
+```shell
+$ kpt fn eval [PKG_DIR] -i set-labels:v0.1.5 --match-annotation foo=bar --exclude-annotation foo=baz
+```
+
+```yaml
+# pipeline of Kptfile
+pipeline:
+  mutators:
+  - image: set-labels:v0.1.5
+    selectors:
+    - annotations:
+        foo: bar
+    exclude:
+    - annotations:
+        foo: baz
+```
+
+## Open Issues/Questions
+
+### How should kpt functions handle meta resources?
+
+One concern about this approach is that in order to exclude the Kptfile from being modified by horizontal
+transformations like set-namespace and set-labels, the user will have to write an exclusion field for each function.
+Arguably, the user shouldn't have to provide any special syntax for the Kptfile at all, because kpt and 
+kpt functions should be able to handle it correctly. A mitigation for this is that our own horizontal kpt functions
+can exclude the Kptfile within the function logic itself, and there will be no special handling logic in kpt. 
```

Comment:
The SDK should make filtering out these local config objects easy.

Reply:
Yes, `set-namespace` excludes local-config resources inside its logic.

To have SDK filter out resources, we can check the type for Kptfile, but how can we tell which resources are function configs if not using a new annotation nor reusing local-config?

Reply:
To be more specific, in a hydration pipeline if function1 and function2 have different function config `funcitonConfig1` and `functionConfig2 ` correspondingly, the resourceslist for `function1` is.
```
kind: ResourceList
functionConfig: 
    kind: functionConfig1
    ...
items:
- kind: Kptfile
   ...
- kind: functionConfig2
   ...
```
the resourceslist for `function2` is.
```
kind: ResourceList
functionConfig: 
    kind: functionConfig2
    ...
items:
- kind: Kptfile
   ...
- kind: functionConfig1
   ...
```
function1 does not know which input resource is function2's config, and vice versa.

Reply:
In what cases will the user need to skip functionConfig? If they are our own horizontal functions like set-namespace, then setting the namespace of a functionConfig doesn't seem to really change or break any use cases. And if the user really wants to skip these functionConfigs, then they could use `kind`-based exclusion since every function has its own functionConfig kind, like `SetNamespace`. WDYT?